### PR TITLE
[ISSUE-412] Fix redundant platform label updates

### PR DESCRIPTION
**Pull request by safronovD**
### Issue dell/csi-baremetal#412

- fix extra API calls (check labels before updating)
- fix issues with ctx according to reference from comment
- add parallel resources deploying using `golang.org/x/sync/errgroup`

## PR checklist
- [x] Add link to the issue
- [x] Choose PR label
- [ ] New unit tests added
- [ ] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

### CODE REVIEW

File: pkg/csi_deployment.go [hunk 1]
```diff
@@ -14,6 +15,7 @@ import (
 )
 
 type CSIDeployment struct {
+	ctx            context.Context
```
Comment (AnastasiaShemyakinskaya) [inline pkg/csi_deployment.go:18]: I guess it's not really good practise to use context in such way. For reference  https://www.ardanlabs.com/blog/2019/09/context-package-semantics-in-go.html

**Comment by mishoyama** [comment]
@safronovD can we just have some global variable instead showing that this stage has passed instead of executing extra call to k8s API server?

File: pkg/csi_deployment.go (hunk 1 above)
Comment (safronovD) [reply pkg/csi_deployment.go]: Thank you for the article. Removed context from data structs and updated receiving functions

**Comment by safronovD** [comment]
@mishoyama Do you mean check annotations before true and then just skip? 
We can have situation when node is labeled with node-selector and then node-controller has to create new csibmnode, so we won't update our variable in this case (or we need to do one more API call, which is same)

**Review by AnastasiaShemyakinskaya — APPROVED** [review]

**Review by yurkov-anton — APPROVED** [review]

File: pkg/csi_deployment.go [hunk 2]
```diff
@@ -21,62 +22,51 @@ type CSIDeployment struct {
...
-
-	if err := c.controller.Update(csi, scheme); err != nil {
-		return err
-	}
+	errs, ctx := errgroup.WithContext(ctx)
 
-	if err := c.extender.Update(csi, scheme); err != nil {
-		return err
-	}
+	errs.Go(func() error { return c.nodeController.Update(ctx, csi, scheme) })
```
Comment (mishoyama) [inline pkg/csi_deployment.go:54]: Why do you want to deploy components in parallel here? Shouldn't Patcher be run when Extender is deployed only?

Comment (safronovD) [reply pkg/csi_deployment.go:54]: Currently, if we have an error in the stage, following stages after failed one are not started at all. It makes slower deploying process and increases API calls number. 
Extender will wait until Patcher have done its work in binding stage. And if we want to start it after patch, we need to check if Patcher is running and kube-scheduler is restarted. At the moment we creates them only and don't do these checks

Comment (mishoyama) [reply pkg/csi_deployment.go:54]: I guess this is expected behavior when one component failed to install - don't proceed with others. What do you mean by increased number of API calls?

Comment (safronovD) [reply pkg/csi_deployment.go:54]: I mean when there is an error while reconciling (updating here), we don't really process all components only ones before an error and if one more error below appears, controller has to retry reconcile one more time. As for creating, our components are independent from each other, they can wait before others have been ready after deploying.

But I agree that it's better not to try creating new components if one was failed. What order do you propose? Patcher -> Extender -> NodeController -> Controller -> Node?

Comment (mishoyama) [reply pkg/csi_deployment.go:54]: NodeController -> Driver -> Extender/Patcher

Comment (safronovD) [reply pkg/csi_deployment.go]: Done

**Review by mishoyama — APPROVED** [review]

**Pull request merged**